contrib/zenith/file_cache.c

```diff
@@ -0,0 +1,345 @@
+/*
+ *
+ * file_cache.c
+ *
+ *
+ * Portions Copyright (c) 1996-2021, PostgreSQL Global Development Group
+ * Portions Copyright (c) 1994, Regents of the University of California
+ *
+ *
+ * IDENTIFICATION
+ *	  contrib/zenith/file_cache.c
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include <sys/file.h>
+#include <unistd.h>
+#include <fcntl.h>
+
+#include "postgres.h"
+#include "miscadmin.h"
+#include "pgstat.h"
+#include "pagestore_client.h"
+#include "postmaster/bgworker.h"
+#include "storage/relfilenode.h"
+#include "storage/buf_internals.h"
+#include "storage/latch.h"
+#include "storage/ipc.h"
+#include "storage/lwlock.h"
+#include "utils/dynahash.h"
+#include "utils/guc.h"
+#include "storage/fd.h"
+#include <storage/buf_internals.h>
+
+/*
+ * Local file cache is used to temporary store relations pages in local file system.
+ * All blocks of all relations are stored inside one file and addressed using shared hash map.
+ * Currently LRU eviction policy based on L2 list is used as replacement algorithm.
+ * As far as manipulation of L2-list requires global critical section, we are not using partitioned hash.
+ * Also we are using exclusive lock even for read operation because LRU requires relinking element in L2 list.
+ * If this lock become a bottleneck, we can consider other eviction strategies, for example clock algorithm.
+ *
+ * Cache is always reconstructed at node startup, so we do not need to save mapping somewhere and worry about
+ * its consistency.
+ */
+
+/* Local file storage allocation chunk.
+ * Should be power of two. Using larger than page chunks can
+ * 1. Reduce hash-map memory footprint: 8TB database contains billion pages
+ *    and size of hash entry is 40 bytes, so we need 40Gb just for hash map.
+ *    1Mb chunks can reduce hash map size to 320Mb.
+ * 2. Improve access locality, subsequent pages will be allocated together improving seqscan speed
+ */
+#define CHUNK_SIZE 128 /* 1Mb chunk */
+
+typedef struct FileCacheEntry
+{
+	BufferTag key;
+	uint32    offset;
+	uint32    access_count;
+	uint32    bitmap[CHUNK_SIZE/32];
+	/* LRU list */
+	struct FileCacheEntry* next;
+	struct FileCacheEntry* prev;
+} FileCacheEntry;
+
+typedef struct FileCacheControl
+{
+	uint32 size; /* size of cache file in chunks */
+	FileCacheEntry lru;
+} FileCacheControl;
+
```

Comment:
Shouldn't each FileCacheControl entry have an associated lock, so that concurrent read/write operations in the hashmap won't conflict? This seems especially important since the read/write operations into the file are done without holding the (global) lfc_lock;

(unlinking before IO and re-linking after IO is also lock-ish, but is not very well documented and currently critical to prevent race conditions)

Reply:
It is not possible in Postgres to create arbitrary (dynamically increased) number of RW locks.
And event if it was possible, it is too expensive to create RW lock per entry.
Postgres use another approach: sharded hash table. In this case hash is splitted into several partitions, each of them is protected by its own lock.

But in case of local file cache it is problematic, because I also need critical section for LRU. Well, LRU can be replaced with Clock, which doesn't require global lock. But my experiments shows that hash table access time is negligible comparing with actual IO. So even with large number of clients time spent in hash table access and lock contention is relatively small. I didn't see problems with scaling with increasing number of clients.
So may be it will be sometime it will be rewritten to provide better parallelism. But right now I do not see this problem and want to keep implementation as simple as possible.

Reply:
> It is not possible in Postgres to create arbitrary (dynamically increased) number of RW locks.

It is possible nowadays, actually. Since commit ea9df812d8502fff74e7bc37d61bdc7d66d77a7f. No comment on whether it makes sense in this case.